bpo-40217:  Ensure Py_VISIT(Py_TYPE(self)) is always called for PyType_FromSpec types (reverts GH-19414)

pablogsal opened the pull request:
Heap types now always visit the type in tp_traverse. See added docs for details.

This reverts commit 0169d3003be3d072751dd14a5c84748ab63a249f.

<!-- issue-number: [bpo-40217](https://bugs.python.org/issue40217) -->
https://bugs.python.org/issue40217
<!-- /issue-number -->


Automerge-Triggered-By: @encukou

pablogsal committed:
Revert "bpo-40217: Ensure Py_VISIT(Py_TYPE(self)) is always called for PyType_FromSpec types (GH-19414)"

This reverts commit 0169d3003be3d072751dd14a5c84748ab63a249f.

pablogsal committed:
Fix heap types in the stdlib to correctly visit its parent

pablogsal committed:
Add documentation about the change

encukou reviewed (commented):
Are the `type->tp_traverse == (traverseproc)foo_traverse` checks (and the *but not for subclasses* remarks) necessary? AFAIK, the check will always be true for calls that Python itself makes.
It might necessary to do the check in the reproducer I posted on the bug, but only because that module's own call to `base_traverse`.

But it might be that I'm misunderstanding the situation.

encukou committed:
subtype_traverse: Don't visit the type if base is a heap type.

If the base is a heap type, the responsibility to visit
Py_TYPE(self) is delegated to the base's tp_traverse.
